### DISCUSSION TITLE
How to Add a Listener to the Handshake Future of Netty's Client SSL Handler

### DISCUSSION THREAD

**Discussion by htutman** [micronaut-projects/micronaut-core#11232] [Q&A] [answered]
Hi folks,

I'm working on a service that uses mTLS for outbound communication, and I need to log which certificate chain the HTTP client is actually using. In Netty, this should be achievable with something like:

```Java
sslHandler.handshakeFuture().addListener(future -> {
  if (future.isSuccess()) {
    logCertificates(sslHandler.engine().getSession().getLocalCertificates());
  }
});
```

However, in Micronaut, the `sslHandler` is initialized within [`ConnectionManager.java`](https://github.com/micronaut-projects/micronaut-core/blob/4.7.x/http-client/src/main/java/io/micronaut/http/client/netty/ConnectionManager.java), and configured in the private method [`configureSslHandler`](https://github.com/micronaut-projects/micronaut-core/blob/4.7.x/http-client/src/main/java/io/micronaut/http/client/netty/ConnectionManager.java#L637), and I haven't found a way to hook into it.

Is there a recommended approach to achieve this in Micronaut's HTTP client?

Thanks.

**Comment by yawkat** [comment, marked as answer]
you can use a pipeline customizer

**Reply by htutman** [reply]
Thank you @yawkat!
